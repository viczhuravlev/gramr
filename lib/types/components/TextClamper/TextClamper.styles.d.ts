/// <reference types="react" />
import * as T from './TextClamper.types';
export declare const Clamper: import("@emotion/styled-base").StyledComponent<import("react").DetailedHTMLProps<import("react").HTMLAttributes<HTMLSpanElement>, HTMLSpanElement>, T.Clamper, any>;
