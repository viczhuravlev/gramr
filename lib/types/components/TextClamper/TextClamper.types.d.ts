import { ReactNode } from 'react';
export interface Clamper {
    clamp?: string;
}
export interface TextClamperProps extends Clamper {
    className?: string;
    children: ReactNode;
}
