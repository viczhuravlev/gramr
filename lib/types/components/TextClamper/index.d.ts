export { TextClamperProps } from './TextClamper.types';
export { default as TextClamper } from './TextClamper';
