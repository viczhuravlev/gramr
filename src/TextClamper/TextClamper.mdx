---
name: TextClamper
---

import { Playground, PropsTable } from 'docz';

import TextClamper from './TextClamper';

# TextClamper

<PropsTable of={TextClamper} />

Warning! This is the support for WebKit's propriety and undocumented implementation of line clamp.

More info: Line clamps are part of the CSS Overflow Module Level 3 which is currently in Editor's Draft and totally unsupported at the moment.

## Basic usage

<Playground>
  <TextClamper>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
    laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
    voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
    non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </TextClamper>
</Playground>

## Clamp

<Playground>
  <TextClamper clamp="3">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
    labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco
    laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in
    voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat
    non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </TextClamper>
</Playground>
