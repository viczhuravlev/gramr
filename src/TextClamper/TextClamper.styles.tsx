import styled from '@emotion/styled';

import * as T from './TextClamper.types';

export const Clamper = styled.span<T.Clamper>`
  overflow: hidden;
  white-space: normal;
  /* stylelint-disable */
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: ${props => props.clamp || '1'};
  /* stylelint-enable */
`;
