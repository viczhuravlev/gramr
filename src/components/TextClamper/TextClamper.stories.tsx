import React from 'react';
import { storiesOf } from '@storybook/react';

import { TextClamper } from './index';

const warning =
  "Warning! This is the support for WebKit's propriety and undocumented implementation of line clamp.";
const moreInfo =
  "More info: Line clamps are part of the CSS Overflow Module Level 3 which is currently in Editor's Draft and totally unsupported at the moment.";

storiesOf('TextClamper', module)
  .add('Default', () => (
    <>
      <p>
        {warning}
        <br />
        {moreInfo}
      </p>
      <br />
      <TextClamper>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
        commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
        velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint
        occaecat cupidatat non proident, sunt in culpa qui officia deserunt
        mollit anim id est laborum.
      </TextClamper>
    </>
  ))
  .add('Clamp', () => (
    <TextClamper clamp="3">
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
      tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
      veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea
      commodo consequat. Duis aute irure dolor in reprehenderit in voluptate
      velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat
      cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id
      est laborum.
    </TextClamper>
  ));
