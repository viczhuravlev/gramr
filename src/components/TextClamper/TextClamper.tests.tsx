import React from 'react';
import renderer from 'react-test-renderer';

import TextClamper from './TextClamper';

describe('[component] TextClamper', () => {
  test('[Snapshot] TextClamper', () => {
    const tree = renderer
      .create(
        <TextClamper clamp="3">
          splitOnChars (Array). Determines what characters to use to chunk an element into smaller
          pieces. Version 0.1 of Clamp.js would always remove each individual character to check for
          fit. With v0.2, you now have an option to pass a list of characters it can use. For
          example, it you pass an array of ... then it will first remove sentences, then remove
          comma-phrases, and remove words, and finally remove individual characters to try and find
          the correct height. This will lead to increased performance and less looping when removing
          larger pieces of text (such as in paragraphs). The default is set to remove sentences
          (periods), hypens, en-dashes, em-dashes, and finally words (spaces). Removing by character
          is always enabled as the last attempt no matter what is submitted in the array.
        </TextClamper>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
