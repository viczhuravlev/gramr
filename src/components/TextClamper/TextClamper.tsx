import React from 'react';

import * as S from './TextClamper.styles';
import * as T from './TextClamper.types';

function TextClamper(props: T.TextClamperProps): JSX.Element | null {
  const { clamp, children, className } = props;

  if (!children) return null;

  return (
    <S.Clamper className={className} clamp={clamp}>
      {children}
    </S.Clamper>
  );
}

TextClamper.defaultProps = {
  clamp: '1'
};

export default TextClamper;
