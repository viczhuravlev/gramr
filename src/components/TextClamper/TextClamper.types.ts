import { ReactNode } from 'react';

export interface Clamper {
  /** Default: 1 */
  clamp?: string;
}

export interface TextClamperProps extends Clamper {
  className?: string;
  children: ReactNode;
}
