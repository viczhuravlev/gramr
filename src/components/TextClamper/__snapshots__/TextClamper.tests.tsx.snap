// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[component] TextClamper snapShot testing 1`] = `
<span
  className="cb52807"
  style={
    Object {
      "--cb52807-0": "3",
    }
  }
>
  splitOnChars (Array). Determines what characters to use to chunk an element into smaller pieces. Version 0.1 of Clamp.js would always remove each individual character to check for fit. With v0.2, you now have an option to pass a list of characters it can use. For example, it you pass an array of ... then it will first remove sentences, then remove comma-phrases, and remove words, and finally remove individual characters to try and find the correct height. This will lead to increased performance and less looping when removing larger pieces of text (such as in paragraphs). The default is set to remove sentences (periods), hypens, en-dashes, em-dashes, and finally words (spaces). Removing by character is always enabled as the last attempt no matter what is submitted in the array.
</span>
`;
